polkadot/node/network/statement-distribution/src/v2/mod.rs

```diff
@@ -3001,11 +3070,16 @@ pub(crate) fn answer_request(state: &mut State, message: ResponderMessage) {
 
 	// Transform mask with 'OR' semantics into one with 'AND' semantics for the API used
 	// below.
-	let and_mask = StatementFilter {
+	let mut and_mask = StatementFilter {
 		seconded_in_group: !mask.seconded_in_group.clone(),
 		validated_in_group: !mask.validated_in_group.clone(),
 	};
 
+	// Ignore disabled validators when sending the response.
+	let disabled_mask = per_session.disabled_bitmask(confirmed.group_index()).unwrap_or_default();
```

Comment:
This bit is kind of finicky and probably incorrect.

Previously one of the following two invariants held:
1. The request is from the cluster and the response is not required to indicate a quorum
2. The request is not from the cluster and the response is required to indicate quorum

With the introduction of validator disabling, we can have sent a manifest to a peer, then received information about disabling, and then receive a request. As this diff causes us to answer with fewer statements than we observed at the time of broadcasting the manifest, invariant (2) would no longer hold unconditionally for grid requests.

I suggest that
  1. there should be no response at all for grid requests when the backing threshold is no longer met as a result of disabled validators
  2. no request should be made of a validator for a grid candidate when the manifest from the validator does not reach quorum when accounting for disabled validators

This still has some small race conditions but maintains consistency without overtly misbehaving in the protocol and minimizing the distribution of backed-by-malicious candidates.

Reply:
Initially, I've made a change to unwanted_mask to include disabled validators.
However, at the time I wasn't sure if fiddling with the mask is safe and
whether it is used for the same purpose as manifest, to know which statements
a validator already has (it's not) or whether the consistency with the
manifest is ensured (it seems not). So this change was reverted.

> there should be no response at all for grid requests when the backing threshold is no longer met as a result of disabled validators

What about not meeting the threshold even without disabled validators?
After I've made this change, the `local_node_respects_statement_mask` test
started failing. The reason being, the backing threshold is set to 2 and we
only send one statement as a reply. My understanding is that the sender should
not send such requests (filtered by `find_request_target_with_update`), so the
test was unrealistic.